src/Data/Bin/Properties.agda

```diff
@@ -1,143 +1,904 @@
 
-------------------------------------------------------------------------
--- (Bin, _≡_) is a decidable setoid
-
-1#-injective : ∀ {as bs} → as 1# ≡ bs 1# → as ≡ bs
-1#-injective refl = refl
-
-infix 4 _≟_ _≟ₑ_
-
-_≟ₑ_ : ∀ {base} → Decidable (_≡_ {A = Expansion base})
-_≟ₑ_ []       []       = yes refl
-_≟ₑ_ []       (_ ∷ _)  = no λ()
-_≟ₑ_ (_ ∷ _) []        = no λ()
-_≟ₑ_ (x ∷ xs) (y ∷ ys) with x Fin.≟ y | xs ≟ₑ ys
-... | _        | no xs≢ys = no (xs≢ys ∘ proj₂ ∘ ∷-injective)
-... | no  x≢y  | _        = no (x≢y   ∘ proj₁ ∘ ∷-injective)
-... | yes refl | yes refl = yes refl
-
-_≟_ : Decidable {A = Bin} _≡_
-0#    ≟ 0#    = yes refl
-0#    ≟ bs 1# = no λ()
-as 1# ≟ 0#    = no λ()
-as 1# ≟ bs 1# with as ≟ₑ bs
-... | yes refl  = yes refl
-... | no  as≢bs = no (as≢bs ∘ 1#-injective)
-
-≡-isDecEquivalence : IsDecEquivalence _≡_
-≡-isDecEquivalence = record
-  { isEquivalence = isEquivalence
-  ; _≟_           = _≟_
-  }
-
-≡-decSetoid : DecSetoid _ _
-≡-decSetoid = decSetoid _≟_
+open import Data.Nat.Properties as ℕp using (2m≢1+2n; module ≤-Reasoning)
+open import Data.Product    using (_,_; proj₁; ∃)
+open import Data.Sum        using (_⊎_; inj₁; inj₂)
+open import Function        using (_∘_; _$_; id; const; case_of_)
+open import Level           using () renaming (zero to 0ℓ)
+open import Relation.Binary using (Setoid; DecSetoid; IsDecEquivalence)
+open import Relation.Binary.Core using (_←→_)
+open import Relation.Binary.PropositionalEquality as P using
+                 (_≡_; _≢_; _≗_; refl; sym; trans; cong; cong₂; isEquivalence)
+open import Relation.Nullary          using (yes; no)
+open import Relation.Nullary.Negation using (contradiction)
+
+open P.≡-Reasoning
+open ≤-Reasoning using () renaming (begin_ to begin≤_; _∎ to _end≤;
+                                    _≡⟨_⟩_ to _≡≤[_]_; _≤⟨_⟩_ to _≤[_]_)
+
+import Algebra.Properties.CommutativeSemigroup ℕp.+-semigroup ℕp.+-comm
+                                                     as Of+ℕ-semigroup
+import Algebra.Properties.CommutativeSemigroup ℕp.*-semigroup ℕp.*-comm
+                                                     as Of*ℕ-semigroup
+open FuncProp (_≡_ {A = Bin}) using
+              (Associative; Commutative; LeftIdentity; RightIdentity;
+               LeftZero; RightZero; LeftCancellative; RightCancellative;
+               _DistributesOverˡ_; _DistributesOverʳ_; _DistributesOver_)
+
+open Bin
+
+2suc≢0 :  {x : Bin} → 2suc x ≢ 0#
+2suc≢0 ()
+
+suc2*≢0 :  {x : Bin} → suc2* x ≢ 0#
+suc2*≢0 ()
+
+|x|≡0⇒x≡0 :  {x : Bin} → size x ≡ 0 → x ≡ 0#
+|x|≡0⇒x≡0 {0#} _ =  refl
+
+suc≢0 :  ∀ {x} → suc x ≢ 0#
+suc≢0 {0#}      ()
+suc≢0 {2suc _}  ()
+suc≢0 {suc2* _} ()
+
+x≢0⇒x+y≢0 :  {x : Bin} (y : Bin) → x ≢ 0# → x + y ≢ 0#
+x≢0⇒x+y≢0 {2suc _} 0# _   =  λ()
+x≢0⇒x+y≢0 {0#}     _  0≢0 =  contradiction refl 0≢0
+
+2suc-as∘ :  2suc ≗ 2* ∘ suc
+2suc-as∘ 0#        =  refl
+2suc-as∘ (2suc x)  =  begin
+  2suc (2suc x)        ≡⟨ cong 2suc (2suc-as∘ x) ⟩
```

Comment:
To be honest I think the whole reasoning chain can probably be removed given how simple both the property and the proof is.

Reply:
Do you mean to remove the lemma of ``2suc-as∘`` ?
It is used in this module twice, and many times in the modules to appear.
Indeed, its proof looks so as if the lemma can be omitted.
But look, for example, at the next function. It has a proof step
```
suc2*-as∘ (suc2* x) = begin
  ... 
  suc (2suc 2x)       ≡⟨ cong suc (2suc-as∘ 2x) ⟩
                                   -------
  suc (2* (suc 2x))
```
I wonder of what to set here instead of ``2suc-as∘`` (it is not known whether 2x is 0#). 
I do not know how to get rid of ``2suc-as∘``.
Also it has sense for symmetry, because there is ``suc2*-as∘``.

?

Reply:
Done.
(The old code helped the reader to recall the normalization steps of the proof).